common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.text;
+
+import com.google.common.collect.Iterables;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.type.Pair;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.Style;
+import org.apache.commons.lang3.ArrayUtils;
+
+public final class StyledText {
+    private final Component temporaryWorkaround;
+
+    private final List<StyledTextPart> parts;
+
+    private final List<ClickEvent> clickEvents;
+    private final List<HoverEvent> hoverEvents;
+
+    StyledText(Component component) {
```

Comment:
I'm not fond of having complicated constructors. Can't you move this logic into the factory method instead, and when you have finished parsing the component, just call the "thin" StyledText constructor? (And similarly for the String)

Reply:
(This is not just superstition; constructors are specially treated by the JLS, the object is not fully constructed but you can still leak out `this` which can cause very difficult bugs, etc. For a static method, it is verboten to use `this` so that cannot happen, the compiler will make sure you write correct code.)